File: framework/languages/java/samples/java/GameQuest/Server/GameApi/src/main/java/systems/zlink/samples/gamequest/server/gameapi/handlers/UnlockFeatureHandler.java
package systems.zlink.samples.gamequest.server.gameapi.handlers;

import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.gamequest.server.gameapi.application.GameplayActionService;
import systems.zlink.samples.gamequest.shared.contracts.Messages;

@ZLinkHandlerGroup("gameapi")
public final class UnlockFeatureHandler
    implements ZLinkRequestHandler<Messages.UnlockFeatureReq, Messages.UnlockFeatureRes> {
    private final GameplayActionService actions;

    public UnlockFeatureHandler(GameplayActionService actions) {
        this.actions = actions;
    }

    @Override
    public Messages.UnlockFeatureRes handle(Messages.UnlockFeatureReq request, ZLinkRequestContext context) {
        return actions.unlockFeature(request);
    }
}

File: framework/languages/java/samples/kotlin/GameQuest/Server/GameApi/src/main/kotlin/systems/zlink/samples/kotlin/gamequest/server/gameapi/handlers/UnlockFeatureHandler.kt
package systems.zlink.samples.kotlin.gamequest.server.gameapi.handlers

import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.samples.kotlin.gamequest.server.gameapi.application.GameplayActionService
import systems.zlink.samples.kotlin.gamequest.shared.contracts.UnlockFeatureReq
import systems.zlink.samples.kotlin.gamequest.shared.contracts.UnlockFeatureRes

@ZLinkHandlerGroup("gameapi")
class UnlockFeatureHandler(
    private val actions: GameplayActionService,
) : ZLinkSuspendingRequestHandler<UnlockFeatureReq, UnlockFeatureRes> {
    override suspend fun handle(request: UnlockFeatureReq, context: ZLinkRequestContext): UnlockFeatureRes =
        actions.unlockFeature(request)
}
